### PR TITLE
chore(ci): make the fuzz job run, and make it explore HTML

### PR DESCRIPTION
**Pull request by harlan-zw**
### 🔗 Linked issue

Related to #195, related to #200

### ❓ Type of change

- [ ] 📖 Documentation
- [x] 🐞 Bug fix
- [x] 👌 Enhancement
- [ ] ✨ New feature
- [ ] 🧹 Chore
- [ ] ⚠️ Breaking change

### 📚 Description

Two parts: the weekly job never ran, and once it runs it should spend its 300s inside the parser rather than outside it.

#### 1. The job has never passed

Added in #126 on 2026-07-15, three scheduled runs (07-20, 07-27, 08-03), all five targets failing in every one, before fuzzing starts:

```
info: install-action does not support cargo-fuzz (updating install-action might resolve this); fallback to cargo-binstall
error: sanitizer is incompatible with statically linked libc, disable it using `-C target-feature=-crt-static`
error[E0463]: can't find crate for `std`
Error: failed to build fuzz script: ... "cargo" "build" ... "--target" "x86_64-unknown-linux-musl" ...
```

install-action dropped cargo-fuzz, so it falls back to cargo-binstall, which fetches the musl-linked build. cargo-fuzz defaults `--target` to its own build triple, so the fuzzers were being built for `x86_64-unknown-linux-musl`, where the sanitizer cannot link against a static libc.

- pin `--target x86_64-unknown-linux-gnu`, which holds regardless of how cargo-fuzz was installed
- add `rust-src` to the nightly toolchain, in case cargo-fuzz passes `-Zbuild-std`
- cache `corpus/` per target. It is gitignored, so every weekly run was starting cold

#### 2. An HTML preset for the fuzzers

The byte-oriented targets have to rediscover HTML from random bytes, so most of a run is spent on inputs the tokenizer rejects in its first branch. Two additions:

- **`fuzz/html.dict`**, a libFuzzer dictionary of tags, attributes, entities and malformed fragments (`<div class="x`, `</ div>`, `<!-- unterminated`, `&#xZZ;`). Passed to every target; tokens are spliced into the raw input, so the `Arbitrary`-derived targets get it through their `String` fields too.
- **`fuzz_html_grammar`**, a structure-aware target that emits a flat token stream (open, close, self-closing, void, stray `<`, unclosed attribute, `=` with no value) over a vocabulary chosen for handlers that behave differently: raw-text elements, void elements, table and list structure, unknown tags, case variants. Because tokens are emitted flat rather than as a tree, unbalanced and mismatched markup falls out on its own. It runs each document through both the one-shot and the streaming path at a fuzzer-chosen chunk width.

The byte-oriented targets stay. A grammar only produces what it knows about, so it cannot replace random bytes, only spend more time past the tokenizer.

**Measured on cold corpora, 90s each, same machine:**

| setup | runs | cov (edges) | ft (features) |
|---|---|---|---|
| `fuzz_html_to_markdown`, no dict | 2,498,451 | 3,650 | 15,701 |
| `fuzz_html_to_markdown`, with dict | 1,875,977 | 4,954 (+36%) | 20,514 (+31%) |
| `fuzz_html_grammar` | 935,812 | 5,784 (+58%) | 24,657 (+57%) |

The grammar target reaches the most coverage on a third of the executions.

**It also finds real bugs.** Run against `main` for 420s, `fuzz_html_grammar` independently rediscovered the panic #200 fixes:

```
thread panicked at crates/core/src/convert/output.rs:896:
start byte index 27 is out of bounds of `*emphasis* _under_ `tick``
```

Confirmed clean on #200's branch. No crash input is committed here, since #200 already carries a seed for that bug.

**Merge order:** independent of #200, but both touch `fuzz.yml` and `fuzz/Cargo.toml`, so whichever lands second needs a trivial rebase (the matrix gains both `fuzz_html_grammar` and `fuzz_streaming_options`).

Worth a manual `workflow_dispatch` once this lands rather than waiting for Monday, since the musl half of the failure cannot be reproduced locally.

## Summary by CodeRabbit

* **Tests**
  * Added expanded automated fuzz testing for HTML and Markdown conversion.
  * Coverage now includes malformed markup, streaming conversion, sanitization, URL handling, plugins, and edge-case HTML inputs.
  * Added reusable HTML test data to improve detection of parsing and conversion issues.

### CODE REVIEW

File: .github/workflows/fuzz.yml [hunk 1]
```diff
@@ -37,9 +40,22 @@ jobs:
...
           workspaces: crates/core/fuzz -> target
 
+      # corpus/ is gitignored, so without this every weekly run starts cold and
+      # rediscovers the same shallow inputs for 300s
+      - name: Restore corpus
+        uses: actions/cache@0057852bfaa89a56745cba8c7296529d2fc39830 # v4.3.0
+        with:
+          path: crates/core/fuzz/corpus/${{ matrix.target }}
+          key: fuzz-corpus-${{ matrix.target }}-${{ github.run_id }}
+          restore-keys: fuzz-corpus-${{ matrix.target }}-
```
[the remark this answers is not in the document]
Comment (harlan-zw) [reply .github/workflows/fuzz.yml:51]: Fixed in de64d1c's follow-up: split into `actions/cache/restore` + `actions/cache/save` with `if: always()`, and the key now includes `github.run_attempt` so a rerun does not hit the previous attempt's immutable entry. Good catch, a crashing run is precisely the one whose corpus is worth keeping.
